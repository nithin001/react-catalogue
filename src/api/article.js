import instance from './base';
import Ajv from 'ajv';

export const get = (articleId) => {
  return new Promise((resolve, reject) => {
    instance.get(`/article/${articleId}`)
      .then((response) => {
        const data = response.data;
        var ajv = new Ajv();
        var validate = ajv.compile(schema);
        var valid = validate(data);
        if (valid) {
          resolve(data);
        } else {
          reject('backend_error');
        }
      })
      .catch(error => {
        if (error.response && error.response.status === 404) {
          reject('not_found');
          return;
        }
        reject('backend_error');
      });
    ;
  });
};

const schema = {
  '$id': 'http://challenge.monoqi.net/article/{articleId}',
  'type': 'object',
  'definitions': {},
  '$schema': 'http://json-schema.org/draft-07/schema#',
  'additionalProperties': false,
  'properties': {
    'sku': {
      '$id': '/properties/sku',
      'type': 'string',
      'title': 'The Sku Schema ',
      'default': ''
    },
    'name': {
      '$id': '/properties/name',
      'type': 'string',
      'title': 'The Name Schema ',
      'default': ''
    },
    'image': {
      '$id': '/properties/image',
      'type': 'string',
      'title': 'The Image Schema ',
      'default': ''
    },
    'description': {
      '$id': '/properties/description',
      'type': 'string',
      'title': 'The Description Schema ',
      'default': ''
    },
    'price': {
      '$id': '/properties/price',
      'type': 'object',
      'additionalProperties': false,
      'properties': {
        'amount': {
          '$id': '/properties/price/properties/amount',
          'type': 'number',
          'title': 'The Amount Schema ',
          'default': 0
        },
        'currency': {
          '$id': '/properties/price/properties/currency',
          'type': 'string',
          'title': 'The Currency Schema ',
          'default': ''
        }
      },
      'required': [
        'amount',
        'currency'
      ]
    }
  },
  'required': [
    'sku',
    'name',
    'image',
    'description',
    'price'
  ]
};