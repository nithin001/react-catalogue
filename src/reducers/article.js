import Immutable from 'immutable';
const catalog = (state = Immutable.Map(), action) => {
  switch (action.type) {
    case 'POPULATE_ARTICLE': {
      const articles = action.payload;
      return Immutable.fromJS(articles);
    }
  }
  return state;
};
export default catalog;