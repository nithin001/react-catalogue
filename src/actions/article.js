import { get } from '../api/article';
export const load = (articleId) => {
  return (dispatch) => {
    dispatch({ type: 'LOAD_ARTICLE_START' });
    return get(articleId)
      .then((article) => {
        dispatch({ type: 'POPULATE_ARTICLE', payload: article });
        dispatch({ type: 'LOAD_ARTICLE_COMPLETE' });
      })
      .catch(() => {
        dispatch({ type: 'LOAD_ERROR' });
      });
  };
};