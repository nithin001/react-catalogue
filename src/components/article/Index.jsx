import React from 'react';
import { connect } from 'react-redux';
import { addToCart } from '../../actions/cart';

export const Article = ({ article, addToCart }) => {
  return <div className={'article'}>
    <span className={'name'}>{article.get('name')}</span>
    <span className={'description'}>{article.get('description')}</span>
    <img src={article.get('image')} className={'image'}/>
    <span className={'price-amount'}>{article.getIn(['price', 'amount'])}</span>
    <span className={'price-currency'}>{article.getIn(['price', 'currency'])}</span>
    <button className={'add-to-cart'} onClick={() => {
      addToCart(article.get('sku'));
    }}>Add to cart
    </button>
  </div>;
};

export const mapStateToProps = (state) => {
  return { article: state.article };
};

export const mapDispatchToProp = (dispatch) => ({
    addToCart: (sku) => dispatch(addToCart(sku)),
  }
);

export default connect(mapStateToProps, mapDispatchToProp)(Article);
