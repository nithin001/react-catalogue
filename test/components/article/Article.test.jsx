import React from 'react';
import renderer from 'react-test-renderer';
import article from '../../../src/reducers/article';
import { Article } from '../../../src/components/article';
const mockArticle = {
  'sku': '199203',
  'name': 'IPad',
  'image': 'https://picsum.photos/480/480/?image=6',
  'description': 'test',
  'price': { 'amount': 229, 'currency': 'EUR' }
};
const articleState = article(undefined, { type: 'POPULATE_ARTICLE', payload: mockArticle });
describe('Article component', () => {
  it('should render the article item with name, image, description, price', () => {
    const props = {
      article: articleState
    };
    const article = renderer.create(<Article
      {...props}
    />);
    expect(article).toMatchSnapshot();
  });
});