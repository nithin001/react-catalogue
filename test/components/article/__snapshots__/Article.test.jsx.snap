// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article component should render the article item with name, image, description, price 1`] = `
<div
  className="article"
>
  <span
    className="name"
  >
    IPad
  </span>
  <span
    className="description"
  >
    test
  </span>
  <img
    className="image"
    src="https://picsum.photos/480/480/?image=6"
  />
  <span
    className="price-amount"
  >
    229
  </span>
  <span
    className="price-currency"
  >
    EUR
  </span>
  <button
    className="add-to-cart"
    onClick={[Function]}
  >
    Add to cart
  </button>
</div>
`;
